comp/serializer/compression/impl-gzip/gzip_strategy.go

```diff
@@ -0,0 +1,139 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package compressionimpl provides a set of functions for compressing with zlib / zstd / gzip
+package gzipimpl
+
+import (
+	"bytes"
+	"compress/gzip"
+	"io"
+
+	compression "github.com/DataDog/datadog-agent/comp/serializer/compression/def"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// Requires contains the compression level for gzip compression
+type Requires struct {
+	Level   int
+	NewKind func(string, int) compression.Component
+}
+
+// Provides contains the compression component
+type Provides struct {
+	Comp compression.Component
+}
+
+// GzipStrategy is the strategy for when serializer_compression_kind is gzip
+type GzipStrategy struct {
+	level   int
+	newKind func(string, int) compression.Component
+}
+
+// NewComponent returns a new GzipStrategy
+func NewComponent(req Requires) Provides {
+	level := req.Level
+	if level < gzip.NoCompression {
+		log.Warnf("Gzip log level set to %d, minimum is %d.", level, gzip.NoCompression)
+		level = gzip.NoCompression
+	} else if level > gzip.BestCompression {
+		log.Warnf("Gzip log level set to %d, maximum is %d.", level, gzip.BestCompression)
+		level = gzip.BestCompression
+	}
+
+	return Provides{
+		Comp: &GzipStrategy{
+			level:   level,
+			newKind: req.NewKind,
+		},
+	}
+}
+
+// Compress will compress the data with gzip
+func (s *GzipStrategy) Compress(src []byte) (result []byte, err error) {
+	var compressedPayload bytes.Buffer
+	gzipWriter, err := gzip.NewWriterLevel(&compressedPayload, s.level)
+	defer func() {
+		err = gzipWriter.Close()
+		if err != nil {
```

Comment:
Any reason to not print out the error, here? Too verbose maybe?

Reply:
Previously this `Close` wasn't in a `defer`. I think actually having the `Close` in a `defer` means that you'll miss the footer when serializing the bytes below.

Ref: https://github.com/DataDog/datadog-agent/blob/0218309e5e99add9b3bce0e7210defb7e448e2c5/pkg/logs/sender/content_encoding.go#L54-L73

Reply:
Not sure I understand.. What do you mean by footer?

Reply:
Ah, sorry, the gzip footer. `Close()` writes it out as mentioned by https://pkg.go.dev/compress/gzip#Writer.Close but you've already called `Bytes()` to grab a slice of the written bytes and so this won't include the footer.

Reply:
Oooh.. Excellent catch! I assumed `Flush` would have done that. It looks like `Close` doesn't actually do anything that we would need to do if we return early due to an error. 

So  yes, this should not be in a `defer`.